### ISSUE TITLE
Service/Topic delays or doesn't executing

### ISSUE THREAD

**Issue by rmfrazer** [ros-gopigo/ros-node-ropigo#5]
i Ubuntumate 16.04.3 LTS 
RPi2b 
ROS Kinetic 
Gopigo Firmware Version: 16, 
GoPiGo Board Version: 14

So the services and topic in the script below were showing successful but not executing the hardware calls(This is running everything on the robot). Once I added sleeps into the fuctions I started seeing correct executions of the services and topic MOST of the time. I suspect that this issue has something to do with callbacks/loops/queues. What is the best was to tackle this issue so I can remove the sleep commands and get 100% success on the returns?

**Comment by christianrauch** [comment]
Which script do you mean? Do you have an example program?

**Comment by rmfrazer** [comment]
Hi Christian,
   
  Thanks for the quick response. When I run the ropigo.cpp script. If I add a .5 sec delay at the beginning of a service function it appears to work correctly 75-80% of the time. Without the delay I always get a success return but the hardware call never works.

**Comment by christianrauch** [comment]
The `ropigo.cpp` is just a ROS node as an interface to the GoPiGo hardware. How are you calling the services and reading/writing from/to the topics?

Which service are you referring to? There are two services for dis- and enabling of encoders, and 4 services for dis- and enabling the left and right LED. Can you maybe provide a ROS node as python script (using rospy) to trace the problems you have? At the moment I need to understand what exactly you are trying to do, so that I can try this on my GoPiGo.

Have you also tried to do the same without the ROS node, but by using the GoPiGo C library (https://github.com/DexterInd/GoPiGo/tree/master/Software/C) directly? Maybe you are trying to send commands (Encoder on/off, LED on/off) too fast to the GoPiGo.

**Comment by rmfrazer** [comment]
The ROS node is what I'm wanting to learn. 3 terminals (node,roscore,calls to service). SSHing to the robot on all 3. CPU and memory look fine. 

When launching the node these lines of code respond correctly everytime:
<snippet>
int main(int argc, char **argv) {

    init();

    ROS_INFO("GoPiGo Firmware Version: %d",fw_ver());
    ROS_INFO("GoPiGo Board Version: %d",brd_rev());

    led_on(0);
    usleep(500000);
    led_on(1);
    usleep(500000);
    led_off(0);
    usleep(500000);
    led_off(1);
</snippet>

Example without sleep delay. DEBUG shows no errors.This way never works as far as the hardware successes.

rosservice call /led_on_left - Doesn't enable LED
status: 1
rosservice call /led_on_right - Doesn't enable LED
status: 1
rosservice call /led_on_left - Doesn't enable LED
status: 1
rosservice call /led_on_right - Doesn't enable LED
status: 1

Example with sleep delay at the beginning of each function. This way is random. Works about 75% of the time.Even when I wait between calls its still random as far as if the LED enabling/disabling.
rosservice call /led_on_left - Does enable LED
status: 1
rosservice call /led_on_right - Does enable LED
status: 1
rosservice call /led_off_left - Does disable LED
status: 1
rosservice call /led_on_right - Doesn't disable LED
status: 1
rosservice call /led_on_right - Doesn't disable LED
status: 1
rosservice call /led_on_right - Does disable LED
status: 1

**Comment by christianrauch** [comment]
The `usleep(500000)` at the beginning of the node is just for de-/activating at different times, so you see that the node correctly started. It is not needed for turning the LEDs on or off. If you remove the `usleep`, you will just see a very short flashing.

After launching the ROS node via `roslaunch ropigo ropigo.launch`, I see the LEDs correctly flashing. When calling `rosservice call /led_on_right`in another shell, it correctly activates the right LED and `rosservice call /led_off_right` correctly turns the LED off again.
At the moment, I cannot retrace your issues.
Are you calling the different services only once or quickly one after another?

Do you have code for the workaround, that you are using?

**Comment by rmfrazer** [comment]
I knew what the sleep does :) Here is a snippet that works about 75% of the time. I call the service once and pause before I call the next service. Ill do a left on then right on then right off and left off. If the LED doesn't respond to which ever call I'll wait a few seconds and call it again. Sometime works and sometime I have to call it more time before it reacts. Its pretty random but if I take the usleep() out it doesn't work at all.

bool led_enable_left(ropigo::SimpleWrite::Request &/*req*/, ropigo::SimpleWrite::Response &res) {
   usleep(500000)
    int ret = led_on(1);
    res.status = int8_t(ret);
    if(ret!=1) ROS_WARN("Error enabling LED left!");
    else ROS_INFO("Left LED enabled");
    return ret==1;
}

bool led_enable_right(ropigo::SimpleWrite::Request &/*req*/, ropigo::SimpleWrite::Response &res) {
   usleep(500000)
    int ret = led_on(0);
    res.status = int8_t(ret);
    if(ret!=1) ROS_WARN("Error enabling LED right!");
    else ROS_INFO("Right LED enabled");
    return ret==1;
}

bool led_disable_left(ropigo::SimpleWrite::Request &/*req*/, ropigo::SimpleWrite::Response &res) {
   usleep(500000)
    int ret = led_off(1);
    res.status = int8_t(ret);
    if(ret!=1) ROS_WARN("Error disabling LED left!");
    else ROS_INFO("Left LED disabled");
    return ret==1;
}

bool led_disable_right(ropigo::SimpleWrite::Request &/*req*/, ropigo::SimpleWrite::Response &res) {
   usleep(500000)
     int ret = led_off(0);
    res.status = int8_t(ret);
    if(ret!=1) ROS_WARN("Error disabling LED right!");
    else ROS_INFO("Right LED disabled");
    return ret==1;
}

**Comment by christianrauch** [comment]
I updated the services and added topics for LED control. These changes are currently on the `dev` branch. I also created example nodes (https://github.com/ros-gopigo/gopigo_examples) for communicating with the GoPiGo node.

Try the following:

1. checkout the `dev` branch of the GoPiGo node
2. download the [gopigo_examples](https://github.com/ros-gopigo/gopigo_examples) into the same workspace
3. recompile the workspace: `catkin build`
4. in one shell on the Raspberry Pi, run the GoPiGo node: `roslaunch ropigo ropigo.launch`
5. in another shell, try the service example: `rosrun gopigo_examples led_client.py`
    - you should see messages printed on the screen, showing that LEDs are enabled and disabled
6. de-/activate the right LED by publishing on topics:
    - on: `rostopic pub /led_right std_msgs/Bool "data: true" -1`
    - off: `rostopic pub /led_right std_msgs/Bool "data: false" -1`

I hope that either the services or the publishing works for you.

**Comment by rmfrazer** [comment]
First Christian thanks for looking into this. I tried the examples multiple times.Output below and no LED light.
pi@pi-ubuntumate:~/gopigoex_ws$ rosrun gopigo_examples led_client.py
Left LED enabled
Left LED disabled
Right LED enabled
Right LED disabled

 Then added the delay in the functions again and the first time I ran led_client.py only one LED lit but the second time I ran it,after restarting the node, both lit. I see where you added delays in the code but so far it looks like they need to be in the function (still not 100% consistent even when I adjust the delay time)and I’m not sure why.

Publishing to the topic didn't work at all.

**Comment by christianrauch** [comment]
Your service calls are successful (otherwise you would receive "Error enabling..."), which means that also the driver calls with [`led_toggle`](https://github.com/DexterInd/GoPiGo/blob/05b6c7cb12e29e131a4d4392992f1a22802391cd/Software/C/gopigo.c#L340) are successful. I guess the issues you have are related to the GoPiGo driver.

The delays in the python examples are just for keeping the LEDs active for some time, such that they can be seen. They are not intended for fixing the issues you have.

Since the behaviour you describe is non-deterministic and I cannot repeat it in my setup, I am not sure how to help you further. Is there specific value of the delay, at which de-/activating the LEDs works reliable?

Are you by any chance writing with multiple programs or threads to the GoPiGo at the same time?

Also, have you tried the example program ([basic_test_all.c](https://github.com/DexterInd/GoPiGo/blob/master/Software/C/basic_test_all.c)) that comes with the GoPiGo C library?

**Comment by christianrauch** [comment]
I included mutexes for each read and write operation. I thought this might help in case where the nodes main loop and the services are running at the same time.
Can you checkout the `dev` branch again and try the `led_client.py` example again?

**Comment by rmfrazer** [comment]
I tried the new one and still no go. I will continue to work on this. it took me a while just to figure out the delay in the function would work 75% of the time. Thanks for the quick responses and working on this. I actually built a Pi drone with Erle and ROS and it works well. I wanted to start showing my youngest ROS so I decided to start working with gopigo again. I think the gopigo is a good bot to start learning ROS on so I was glad to find this project. When I figure out the issue I'll update this thread. Thanks again!

"Also, have you tried the example program (basic_test_all.c) that comes with the GoPiGo C library?" - Yes this shows no errors and hardware responds without issue.
